### PR TITLE
Website translation handler

### PR DESCRIPTION
**Pull request by share-with-me**

### CODE REVIEW

**Comment by share-with-me** [comment]
@sushain97 , Travis now succeeds, also I modified to raise any exception that rises. Both front and backend are ready to be reviewed.
#54 too, addressed in this PR!

**Comment by sushain97** [comment]
Sounds good! I’ll try my best to take a look at it all tonight.

**Comment by share-with-me** [comment]
Awesome! I'll get started on spellchecker till then!

File: servlet.py [hunk 1]
```diff
@@ -587,20 +594,107 @@ def urlRepl(self, base, attr, quote, aurl):
...
+        """Cache translated text for a pair and url to memory, and disk.
+        Also caches origtext to disk; see cachePath."""
+        if pair not in self.url_cache:
+            self.url_cache[pair] = {}
+        elif len(self.url_cache[pair]) > self.max_inmemory_url_cache:
+            self.url_cache[pair] = {}
+        ts = time.strftime('%a, %d %b %Y %H:%M:%S GMT', time.gmtime(time.time()))
+        self.url_cache[pair][url] = (ts, translated)
+        if self.url_cache_path is None:
+            logging.info("No --url-cache-path, not storing cached url to disk")
```
Comment (sushain97) [inline servlet.py]: @unhammer I'm assuming that this code came from your branch. Why not just use https://docs.python.org/3/library/tempfile.html?

File: servlet.py [hunk 2]
```diff
@@ -1122,7 +1264,6 @@ def setupHandler(
     Handler.scaleMtLogs = scaleMtLogs
     Handler.verbosity = verbosity
     Handler.doc_pipe_sem = Semaphore(max_doc_pipes)
-
```
Comment (sushain97) [inline servlet.py]: why remove this?

Comment (share-with-me) [inline servlet.py]: Damn! I was experimenting what handlers are doing and removed it by mistake. Added again.

File: servlet.py (hunk 1 above)
Comment (unhammer) [inline servlet.py]: This is not a tempfile, we want it stored between restarts. The jorgal apy server often gets restarted many times a day (they keep updating pairs there), and url translation is slow, so a persistent cache is necessary for usable translation. (apertium.org also gets restarted each night I think).

Comment (sushain97) [inline servlet.py]: Ah, I see!

File: servlet.py [hunk 3]
```diff
@@ -587,43 +594,173 @@ def urlRepl(self, base, attr, quote, aurl):
...
-                                             request_timeout=20.0)
-            response = yield http_client.fetch(request)
-            toTranslate = self.htmlToText(response.body, url)
+        if pair is None:
+            return
+        self.notePairUsage(pair)
+        mode_path = self.pairs['%s-%s' % pair]
+        markUnknown = self.get_argument('markUnknown', default='yes') in ['yes', 'true', '1']
+        url = self.get_argument('url')
+        if url.find('http://') == -1:
```
Comment (sushain97) [inline servlet.py]: This does not work because then you'll end up adding http:// to sites that start with https://.

Comment (share-with-me) [inline servlet.py]: Oh yes. I'd make check like `if url.find('http://') == -1 and url.find('https://') == -1:`

Comment (sushain97) [inline servlet.py]: Why are you using `find` when `startswith` exists?

Comment (share-with-me) [inline servlet.py]: Ohh yeah. Sorry for that. I could use `if not url.startswith('http'):`. If a string doesn't start with http, it obviously doesn't start with https. Does this make a valid check?

Comment (sushain97) [inline servlet.py]: Yeah.

**Pull request merged**